File: medium/133. Clone Graph/133. Clone Graph.py
"""
# Definition for a Node.
"""


class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


from typing import Optional


class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        if not node:
            return node

        # new_node = Node()
        # new_node.val = node.val
        # if not node.neighbors:
        #     return new_node
        # # visited = set()
        # nodes_track = {}
        # queue: Deque[Node] = deque([node])
        # nodes_track[node] = new_node
        # while queue:
        #     item = queue.popleft()
        #     neighbors = item.neighbors
        #     for neighbor in neighbors:
        #         if neighbor not in nodes_track:
        #             new_node = Node()
        #             new_node.val = neighbor.val
        #             nodes_track[neighbor] = new_node
        #             queue.append(neighbor)
        #         nodes_track[item].neighbors.append(nodes_track[neighbor])
        # return nodes_track[node]

        visited = {}

        def clone(curr: Node) -> Optional["Node"]:
            if not curr:
                return None

            if curr.val in visited:
                return visited[curr.val]

            copy = Node(curr.val)
            visited[curr.val] = copy

            for neighbor in curr.neighbors:
                copy.neighbors.append(clone(neighbor))

            return copy

        return clone(node)

File: medium/133. Clone Graph/133. Clone Graph.ts
/**
 * Definition for _Node.
 */

class _Node {
    val: number
    neighbors: _Node[]

    constructor(val?: number, neighbors?: _Node[]) {
        this.val = (val === undefined ? 0 : val)
        this.neighbors = (neighbors === undefined ? [] : neighbors)
    }
}


function cloneGraph(node: _Node | null): _Node | null {
    if (!node)
        return null

    const visited = {}

    const clone = (curr: _Node): _Node => {
        if (curr.val in visited)
            return visited[curr.val]

        const copy = new _Node(curr.val)
        visited[curr.val] = copy

        for (const neighbor of curr.neighbors) {
            copy.neighbors.push(clone(neighbor))
        }

        return copy
    }

    return clone(node)
};